text/0000-compiler-team-rolling-leads.md

```diff
@@ -0,0 +1,282 @@
+Friday T-compiler meetings." (A record of leadership on one or more project
+groups or working groups is probably a good thing to have as well, but is not a
+strict requirement.)
+
+## Term Length and Leader Selection
+
+> *"Choose your leaders with wisdom and forethought.*
+> *To be led by a coward is to be controlled by all that the coward fears.*
+> *To be led by a fool is to be led by the opportunists who control the fool.*
+> *To be led by a thief is to offer up your most precious treasures to be stolen.*
+> *To be led by a liar is to ask to be told lies.*
+> *To be led by a tyrant is to sell yourself and those you love into slavery."*
+> -- Octavia E. Butler
+
+The Rust Compiler Team uses rolling co-leadership model of governance, with a
+senior lead and a junior lead. After serving in their positions for one to two
+years, the leaders, with input from the team, select a teammate who is not a
+current lead, and that teammate becomes the new junior lead. The old junior lead
+becomes the new senior lead, and the old senior lead is again a normal compiler
+team member.
+
+The specific term length is left variable since the timing for when a shift in
+leadership makes sense will depend on context.
+
+In code:
+
+```
+enum Level { Senior, Junior }
+struct Member { lead: Option<Level>, ... }
+
+fn roll(curr_senior: &mut Member, curr_junior: &mut Member, incoming: &mut Member) {
+    assert_eq!(curr_senior.lead, Some(Level::Senior));
+    assert_eq!(curr_junior.lead, Some(Level::Junior));
+    assert_ne!(curr_senior, incoming);
+    assert_ne!(curr_junior, incoming);
+    curr_senior.lead = None;
+    curr_junior.lead = Some(Level::Senior);
+    incoming.lead = Some(Level::Junior);
+}
+```
+
+### Selection process
+
+When the senior co-lead decides that they are ready to step down, and have
+confirmed that the junior co-lead feels ready to take on the senior co-lead
+role, then the two tell the T-compiler team privately about the intention to
+have a rollover in leadership.
+
+Then the T-compiler team members can nominate their teammates to serve as the
+new junior co-lead. We here follow the model of our FCP process: The leads
+should provide a ten-day window for nominations to come in, unless they get
+confirmation that the set of nominations is complete.
+
+After nomination is completed, the outgoing senior and junior co-leads discuss
+the set of nominees, and also, if desired, have short discussions with the
+nominees. Then, the senior and junior co-leads select the new co-lead from the
+set of nominees. And that’s it! Then current leads just need to publish a blog
+post saying that the leadership is scheduled to roll over, who the new junior
+lead is, and the date that it takes effect.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+
+Committing to specific term lengths puts pressure on the leads to identify new
+leaders earlier than they might otherwise.
+
```

Comment:
Relatedly, building up leaders (especially from a finite group where trust is built over time) is a non-trivial process, and not guaranteed to work. There are several reasons that a future leader might be hard to find:
* Joining the compiler team is not pre-requisite on committing to one day take on a leadership role.
* Leadership roles are not for everyone, and that is a feature not a bug.
* Compiler team membership does not change that quickly meaning the pool of possible leaders does not change quickly.

I think it's reasonable to think that there may be situations in the future when no junior candidate is obvious. What happens in that situation?

Reply:
I agree we cannot force every team member to take on the leadership role.

I think the answer here is that the leadership, both jr and sr, need to be mentoring their teammates, and fellow community members, as time passes, in order to ensure that we *do* have a pool of canididates when the time comes.

You're correct that the next generation of leaders won't pop up magically.

Perhaps the most important piece here is this: While we cannot ensure that we always have a member of the compiler team who is ready to pick up the mantle, I think it is safe to say that we should always have some team member **or contributor** who seems well-suited to a leadership role. Thus, part of the job of the T-compiler leads is to idenitfy such talent, and if it includes contributors who are not yet a full team member, then the T-compiler leads should work to figure out what gaps in the experience of those contributors need to be filled before we would be comfortable promoting them to full team members.

The process I am alluding to above is indeed not a short process at all. It may be worthwhile to spell it out somewhere, though I'm not sure if this RFC is itself the right place for it...

Reply:
You asked a more specific question than what I addressed above, let me try again.

> I think it's reasonable to think that there may be situations in the future when no junior candidate is obvious. What happens in that situation?

I think if we truly hit a point where no junior candidate is present, then we will have hit a situation where we will need to ask the current leadership to stay on board for longer than expected. But part of the leads' goals at that point *has* to be proactive seeking of the next generation of leaders.

Asking people to stay in the role of leader ad infinitum because no one else is presenting themselves is unhealthy and not sustainable. So there needs to be an active resposne to the scenario that you describe.

Reply:
Hopefully the text added in 23bf9947662b306a1f4675b8e0529ed42d779a8a addresses the point raised here.